### PR TITLE
XN-192: add new HubSpot library for Go

### PR DESCRIPTION
**Pull request by karhidle**
This is a really simple HubSpot library for Go.

For now it only supports:

- API Key authentication 
- Contact Creation

Later we'll probably add support for:
- Update Contact
- Archive Contact

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+BASE_DIR= github.com/teamexos/hubspot-api-go
+BUILD_CMD= env GOOS=linux go build -ldflags="-s -w" -o
+TEST_CMD= go test -timeout 30s
+
```
Comment (cliffom) [inline Makefile:4]:
```suggestion
```

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+BASE_DIR= github.com/teamexos/hubspot-api-go
+BUILD_CMD= env GOOS=linux go build -ldflags="-s -w" -o
```
Comment (cliffom) [inline Makefile:2]: This isn't needed since we aren't building an artifact for output

```suggestion
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+TEST_CMD= go test -timeout 30s
+
+
+.PHONY: build clean deploy
+
+build:
+	dep ensure -v
```
Comment (cliffom) [inline Makefile:8]: Not needed
```suggestion
```

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+clean:
+	rm -rf ./bin ./vendor Gopkg.lock
```
Comment (cliffom) [inline Makefile:11]: Not needed
```suggestion
```

File: README.md [hunk 5]
```diff
@@ -1 +1,50 @@
...
+### Requirements
+
+- Golang (1.14+)
+  - `dep` (0.5.4)
+  
+### Build and Test
+
+```sh
+make build
+make test
```
Comment (cliffom) [inline README.md:14]:
```suggestion
make test
```

File: README.md [hunk 6]
```diff
@@ -1 +1,50 @@
-# hubspot-api-go
\ No newline at end of file
+# hubspot-api-go
+
+v3 SDK(Client) 
+
+### Requirements
+
+- Golang (1.14+)
+  - `dep` (0.5.4)
```
Comment (cliffom) [inline README.md:8]:
```suggestion
- Golang (1.14+)
```

File: hubspot/contact.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// NewContact creates a new Contact Body representation
+// TODO: I'll change this so it's more flexible and allows different properties
+func NewContact(
+	firstName string,
+	lastName string,
+	email string,
+	workEmail string,
+	company string) *Contact {
```
Comment (cliffom) [inline hubspot/contact.go:16]: You addressed my comment in the TODO but I would recommend making a `HubspotContactInput{}` struct for the input as well as a `HubspotContactOutput{}` struct for the return. It is a pattern you see a lot with the AWS SDK.

File: hubspot/client.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+// CreateContact creates a new Contact in HubSpot
+func (c *Client) CreateContact(body Contact) (*Contact, ErrorResponse) {
+	var contact Contact
+	log.Printf("INFO: attempting to create HubSpot Contact")
+
+	requestBody, err := json.Marshal(body)
+	if err != nil {
+		log.Printf("ERROR: could not marshal the provided contact body, err: %v", err)
+		return &contact, ErrorResponse{Status: "error", Message: "invalid contact body"}
```
Comment (cliffom) [inline hubspot/client.go:87]:
```suggestion
		return nil, ErrorResponse{Status: "error", Message: "invalid contact body"}
```

File: hubspot/client.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	}
+	r, err := c.request(
+		fmt.Sprintf("%s/crm/%s/objects/contacts/?hapikey=%s", c.APIBaseURL, c.APIVersion, c.APIKey),
+		http.MethodPost,
+		requestBody)
+
+	if err != nil {
+		log.Printf("ERROR: unable to create hubspot contact")
+		return &contact,
```
Comment (cliffom) [inline hubspot/client.go:96]:
```suggestion
		return nil,
```

File: hubspot/client.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		var errorResponse ErrorResponse
+		err := json.Unmarshal(r.Body, &errorResponse)
+		msg := "ERROR: unable to create HubSpot account. "
+		if err != nil {
+			log.Printf("%sUnable to unmarshall error response.", msg)
+		} else {
+			log.Printf("%sGot error: %v.", msg, errorResponse.Message)
+		}
+		errorResponse.StatusCode = r.StatusCode
+		return &contact, errorResponse
```
Comment (cliffom) [inline hubspot/client.go:110]:
```suggestion
		return nil, errorResponse
```

File: hubspot/client.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	if err := json.Unmarshal(r.Body, &contact); err != nil {
+		msg := fmt.Sprintf("could not unmarshall HubSpot response, err: %v", err)
```
Comment (cliffom) [inline hubspot/client.go:114]:
```suggestion
		msg := fmt.Sprintf("could not unmarshal HubSpot response, err: %v", err)
```

File: README.md [hunk 12]
```diff
@@ -1 +1,49 @@
...
```
Comment (karhidle) [inline README.md:8]:
```suggestion
  - `dep` (0.5.4)
```

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,9 @@
+BASE_DIR= github.com/teamexos/hubspot-api-go
+TEST_CMD= go test -timeout 30s
+
+.PHONY: build clean deploy
```
Comment (cliffom) [inline Makefile:4]:
```suggestion
.PHONY: test
```

File: Makefile [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+
+
+
```
Comment (cliffom) [inline Makefile:7]:
```suggestion
```

File: hubspot/client.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	log.Printf("INFO: request successful, got response status: %v", r.StatusCode)
+
+	// prepare response
+	response.StatusCode = r.StatusCode
+	if err := json.NewDecoder(r.Body).Decode(&response.Body); err != nil {
+		log.Printf("ERROR: could not decode HubSpot response, err: %v", err)
+		return &response, errors.New("ERROR: could not decode response")
+	}
+
+	defer r.Body.Close()
```
Comment (cliffom) [inline hubspot/client.go:159]: If using `defer`, this should be moved right after line 144 where `r` is assigned.

Comment (karhidle) [reply hubspot/client.go]: @cliffom done

**Pull request merged**